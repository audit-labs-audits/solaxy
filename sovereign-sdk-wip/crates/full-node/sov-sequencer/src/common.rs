//! Defines the [`Sequencer`] trait and related types.

use std::collections::HashMap;
use std::fmt::Debug;
use std::future::Future;
use std::sync::Arc;

use async_trait::async_trait;
use axum::http::StatusCode;
use borsh::{BorshDeserialize, BorshSerialize};
use sov_blob_sender::{BlobInternalId, BlobSenderHooks};
use sov_db::ledger_db::LedgerDb;
use sov_modules_api::capabilities::{AuthenticationOutput, TransactionAuthenticator};
use sov_modules_api::rest::utils::ErrorObject;
use sov_modules_api::rest::{ApiState, StateUpdateReceiver};
use sov_modules_api::*;
use sov_modules_stf_blueprint::{PreExecError, Runtime};
use sov_rest_utils::{json_obj, to_json_object};
use sov_rollup_interface::node::da::DaService;
use sov_rollup_interface::node::ledger_api::{ItemOrHash, LedgerStateProvider, QueryMode};
use sov_rollup_interface::node::{future_or_shutdown, FutureOrShutdownOutput};
use tokio::sync::{broadcast, watch, Mutex, RwLock};
use tokio::time::timeout;
use tracing::{info, trace};

use crate::{SequencerEvent, SequencerNotReadyDetails, TxHash, TxStatus, TxStatusManager};

/// The [`Sequencer`] trait is responsible for accepting transactions and
/// assembling them into batches.
#[async_trait]
pub trait Sequencer: Send + Sync + 'static {
    /// What data is returned to clients when a transaction is accepted.
    type Confirmation: serde::Serialize + Send + Sync + 'static;
    /// The rollup spec.
    type Spec: Spec;
    /// The rollup's [`Runtime`].
    type Rt: Runtime<Self::Spec>;
    /// The [`DaService`] used by the node (and sequencer).
    type Da: DaService<Spec = <Self::Spec as Spec>::Da>;

    /// Only available if the [`Sequencer`] supports events streaming.
    async fn subscribe_events(&self) -> Option<broadcast::Receiver<SequencerEvent<Self::Rt>>> {
        None
    }

    /// Get a list of events by event numbers.
    ///
    /// This is only available if the [`Sequencer`] supports events streaming.
    async fn list_events(
        &self,
        _event_numbers: &[u64],
    ) -> Result<
        Vec<RuntimeEventResponse<<Self::Rt as RuntimeEventProcessor>::RuntimeEvent>>,
        anyhow::Error,
    > {
        Ok(vec![])
    }

    /// Returns an [`ApiState`] subscribed to updates of the batch builder's state.
    fn api_state(&self) -> ApiState<Self::Spec>;

    /// Checks whether the batch builder is ready to accept transactions.
    async fn is_ready(&self) -> Result<(), SequencerNotReadyDetails>;

    /// Queries a transaction's status.
    async fn tx_status(
        &self,
        tx_hash: &TxHash,
    ) -> anyhow::Result<TxStatus<<<Self::Spec as Spec>::Da as DaSpec>::TransactionId>>;

    /// Updates the sequencer's view of the state of the rollup.
    async fn update_state(
        &self,
        update_info: StateUpdateInfo<<Self::Spec as Spec>::Storage>,
    ) -> anyhow::Result<()>;

    /// Adds a **not-encoded** transaction to the mempool. The [`Sequencer`]
    /// implementation itself is responsible for "encoding" the transaction.
    ///
    /// Can return an error if transaction is invalid or mempool is full.
    async fn accept_tx(
        &self,
        tx: FullyBakedTx,
    ) -> Result<AcceptedTx<Self::Confirmation>, ErrorObject>;

    /// Can be used to query and update the status of transactions.
    fn tx_status_manager(&self) -> &TxStatusManager<<Self::Spec as Spec>::Da>;
}

/// A transaction that has been accepted by the batch builder.
#[serde_with::serde_as]
#[derive(Debug, Clone, serde::Serialize)]
pub struct AcceptedTx<C> {
    /// Encoded transaction, as will appear on-chain.
    #[serde_as(as = "serde_with::base64::Base64")]
    pub tx: FullyBakedTx,
    /// Hash of the transaction.
    pub tx_hash: TxHash,
    /// Confirmation data. Could be empty, a receipt, or other data.
    pub confirmation: C,
}

impl<C> AcceptedTx<C> {
    /// Maps the inner confirmation data.
    pub fn map_confirmation<D>(self, f: impl FnOnce(C) -> D) -> AcceptedTx<D> {
        AcceptedTx {
            tx: self.tx,
            tx_hash: self.tx_hash,
            confirmation: f(self.confirmation),
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq, BorshSerialize, BorshDeserialize)]
pub struct WithCachedTxHashes<I> {
    pub inner: I,
    pub tx_hashes: Arc<[TxHash]>,
}

/// Sends [`TxStatusManager`] notifications upon blob status changes.
pub struct TxStatusBlobSenderHooks<Da: DaSpec> {
    txsm: TxStatusManager<Da>,
    tx_hashes_by_blob_id: RwLock<HashMap<BlobInternalId, Arc<[TxHash]>>>,
}

impl<Da: DaSpec> TxStatusBlobSenderHooks<Da> {
    pub fn new(txsm: TxStatusManager<Da>) -> Self {
        Self {
            txsm,
            tx_hashes_by_blob_id: Default::default(),
        }
    }

    pub async fn add_txs(&self, blob_id: BlobInternalId, tx_hashes: Arc<[TxHash]>) {
        self.tx_hashes_by_blob_id
            .write()
            .await
            .insert(blob_id, tx_hashes);
    }

    async fn notify_all(&self, blob_id: BlobInternalId, status: TxStatus<Da::TransactionId>) {
        let tx_hashes = self
            .tx_hashes_by_blob_id
            .read()
            .await
            .get(&blob_id)
            .cloned()
            .unwrap_or_default();

        for tx_hash in &*tx_hashes {
            trace!(
                %tx_hash,
                ?status,
                %blob_id,
                "Notifying about transaction status as a result of blob submission",
            );
            self.txsm.notify(*tx_hash, status.clone());
        }
    }
}

#[async_trait]
impl<Da: DaSpec> BlobSenderHooks for TxStatusBlobSenderHooks<Da> {
    type Da = Da;

    async fn on_published_blob(
        &self,
        blob_id: BlobInternalId,
        _blob_hash: [u8; 32],
        da_tx_id: Da::TransactionId,
    ) {
        self.notify_all(blob_id, TxStatus::Published { da_tx_id })
            .await;
    }

    async fn on_finalized_blob(
        &self,
        blob_id: BlobInternalId,
        _blob_hash: [u8; 32],
        _da_tx_id: <Self::Da as DaSpec>::TransactionId,
    ) {
        self.tx_hashes_by_blob_id.write().await.remove(&blob_id);
    }
}

/// Empty transaction confirmation data.
///
/// Serializes as an empty JSON object.
#[derive(Clone, serde::Serialize)]
pub struct EmptyConfirmation {}

type AuthRes<S, Rt, I> = (
    TxScratchpad<S, I>,
    Result<
        (
            AuthenticationOutput<S, <Rt as DispatchCall>::Decodable>,
            BasicGasMeter<S>,
        ),
        PreExecError,
    >,
);

/// Does something -anything- in a loop every time the [`StateUpdateReceiver`]
/// receives a new value. A new trigger is generated with a timeout even if no
/// value is received.
///
/// Automagically handles shutdown and error checking for you.
pub async fn react_to_state_updates<S, Fut>(
    mut state_update_receiver: StateUpdateReceiver<S::Storage>,
    shutdown_receiver: watch::Receiver<()>,
    task_name: &'static str,
    mut closure: impl FnMut(StateUpdateInfo<S::Storage>) -> Fut,
) where
    S: Spec,
    Fut: Future<Output = anyhow::Result<()>>,
{
    loop {
        let changed_with_timeout = timeout(
            std::time::Duration::from_secs(30),
            state_update_receiver.changed(),
        );

        let fut = future_or_shutdown(changed_with_timeout, &shutdown_receiver);
        let FutureOrShutdownOutput::Output(timeout_result) = fut.await else {
            info!(
                task_name,
                "Shutdown signal receiver, exiting sequencer background task"
            );
            break;
        };

        match timeout_result {
            Ok(changed_result) => {
                // Handle actual state update notification
                if let Err(error) = changed_result {
                    tracing::error!(%error, task_name, "Channel notification failed, exiting sequencer background task. This is a bug, please report it");
                    break;
                }

                let info = (*state_update_receiver.borrow()).clone();
                if let Err(err) = closure(info).await {
                    tracing::error!(%err, task_name, "Error inside the sequencer background task's closure; this is a bug, please report it");
                    break;
                }
            }
            Err(_) => {
                // Timeout elapsed. Log a warning then continue waiting. TODO: handle excessive timeouts (#2938).
                tracing::warn!(
                    task_name,
                    "Timeout waiting for state update notification, continuing to wait"
                );
                continue;
            }
        }
    }
}

pub async fn loop_call_update_state<Seq: Sequencer>(
    seq: Arc<Seq>,
    state_update_receiver: StateUpdateReceiver<<Seq::Spec as Spec>::Storage>,
    shutdown_receiver: watch::Receiver<()>,
) {
    react_to_state_updates::<Seq::Spec, _>(
        state_update_receiver,
        shutdown_receiver,
        "loop_call_update_state",
        |info| async {
            if cfg!(debug_assertions) {
                let skip_flag = std::env::var("SOV_TEST_PAUSE_SEQUENCER_UPDATE_STATE");
                if skip_flag == Ok("1".to_string()) {
                    tracing::warn!("skipping state update due to env var flag");
                    Ok(())
                } else {
                    seq.update_state(info).await
                }
            } else {
                seq.update_state(info).await
            }
        },
    )
    .await;
}

#[tracing::instrument(skip_all, level = "trace")]
pub async fn loop_send_tx_notifications<S: Spec, Rt: RuntimeEventProcessor>(
    state_update_receiver: StateUpdateReceiver<S::Storage>,
    shutdown_receiver: watch::Receiver<()>,
    ledger_db: &LedgerDb,
    txsm: &TxStatusManager<S::Da>,
) {
    // `Arc<Mutex<...>>` is, I suspect, overkill here. It's just a workaround
    // around the `FnMut` closure issues I was banging my head against while writing
    // this.
    let latest_processed_slot_number =
        Arc::new(Mutex::new(state_update_receiver.borrow().slot_number));

    react_to_state_updates::<S, _>(state_update_receiver, shutdown_receiver, "loop_send_tx_notifications", move |info| {
        let latest_processed_slot_number = latest_processed_slot_number.clone();
        async move {
            let storage_slot_number = info.slot_number;
            let range = latest_processed_slot_number.lock().await.range_inclusive(storage_slot_number);

            trace!(%storage_slot_number, "Querying slot data from node to notify about transaction status");

            for slot_number in range {
                let slot = ledger_db
                    .get_slot_by_number::<BatchSequencerReceipt<S>, TxReceiptContents<S>, RuntimeEventResponse<Rt::RuntimeEvent>>(
                        slot_number,
                        QueryMode::Full,
                    )
                    .await?
                    .expect("Received slot notification from node, but it's absent in the ledger. This is a bug, please report it");

                for batch in slot.batches.unwrap_or_default().iter() {
                    let ItemOrHash::Full(batch) = batch else {
                        continue;
                    };
                    for tx in batch.txs.as_deref().unwrap_or_default().iter() {
                        let ItemOrHash::Full(tx) = tx else {
                            continue;
                        };

                        let tx_hash = TxHash::new(tx.hash);
                        trace!(%tx_hash, "Notifying about transaction status as a result of node processing a slot");
                        txsm.notify(tx_hash, TxStatus::Processed);
                    }
                }
            }
            *latest_processed_slot_number.lock().await =info.slot_number;

            Ok(())
        }
    })
    .await;
}

pub fn pre_exec_err_to_accept_tx_err(err: PreExecError) -> ErrorObject {
    match err{
        PreExecError::SequencerError(error) => {
            ErrorObject {
                status: StatusCode::SERVICE_UNAVAILABLE,
                title: "The sequencer is currently unavailable; contact the administrator if the problem persists".to_string(),
                details: json_obj!({
                    "message": error.to_string()
                })
            }

        },
        PreExecError::AuthError(error) => {
            // For certain kinds of authentication errors, 401
            // or 403 would be more appropriate. But we'd have
            // to inspect the error contents to determine the
            // most appropriate status code... so 400 will do.
            ErrorObject {
                status: StatusCode::BAD_REQUEST,
                title: "The transaction is invalid".to_string(),
                details: json_obj!({
                    "message": error.to_string()
                })
            }
        },
    }
}

pub fn generic_accept_tx_error(details: impl std::fmt::Debug) -> ErrorObject {
    ErrorObject {
        status: StatusCode::BAD_REQUEST,
        title: "The transaction is invalid".to_string(),
        details: json_obj!({
            "message": format!("{:?}", details)
        }),
    }
}

pub fn error_not_fully_synced(details: SequencerNotReadyDetails) -> ErrorObject {
    let summary = match details {
        SequencerNotReadyDetails::Syncing { .. } => "The node is not fully synced with the DA head",
        SequencerNotReadyDetails::WaitingOnDa { .. } => {
            "The sequencer is waiting for the DA to finalize more blocks"
        }
        SequencerNotReadyDetails::WaitingOnBlobSender { .. } => {
            "The sequencer is waiting for the blob sender to be ready"
        }
    };
    ErrorObject {
        status: StatusCode::SERVICE_UNAVAILABLE,
        title: format!("{summary}; No new transactions can be accepted, try again later"),
        details: to_json_object(details),
    }
}

pub fn tx_auth<S, Rt, I>(
    tx_scratchpad: TxScratchpad<S, I>,
    gas_price: <S::Gas as Gas>::Price,
    baked_tx: &FullyBakedTx,
) -> AuthRes<S, Rt, I>
where
    S: Spec,
    Rt: Runtime<S>,
    I: StateProvider<S>,
{
    let gas_meter = BasicGasMeter::new_with_gas(<S as GasSpec>::max_tx_check_costs(), gas_price);

    let mut pre_exec_ws = tx_scratchpad.to_pre_exec_working_set(gas_meter);

    let auth_res = match Rt::Auth::authenticate(baked_tx, &mut pre_exec_ws) {
        Ok(ok) => ok,
        Err(err) => {
            let tx_scratchpad = pre_exec_ws.to_scratchpad_and_gas_meter().0;
            return (tx_scratchpad, Err(PreExecError::AuthError(err)));
        }
    };

    let auth_res = (auth_res.0, auth_res.1, Rt::wrap_call(auth_res.2));
    let (tx_scratchpad, gas_meter) = pre_exec_ws.to_scratchpad_and_gas_meter();

    (tx_scratchpad, Ok((auth_res, gas_meter)))
}

/// Checks if the sender of a call is allowed to submit the message via the sequencer.
///
/// Returns true if either...
/// 1. The message is `sequencer_safe`, meaning that submitting will never change the sequencer's config
/// 2. The sender is an admin for this sequencer
pub fn sender_is_allowed<RT: Runtime<S>, S: Spec>(
    runtime: &RT,
    call: &<RT as DispatchCall>::Decodable,
    sender: &S::Address,
    sequencer_address: &<S::Da as DaSpec>::Address,
    admins: &[S::Address],
) -> bool {
    let destination_module = <RT as DispatchCall>::module_info(runtime, call.discriminant());
    destination_module.is_safe_for_sequencer(call.contents(), sequencer_address)
        || admins.contains(sender)
}
