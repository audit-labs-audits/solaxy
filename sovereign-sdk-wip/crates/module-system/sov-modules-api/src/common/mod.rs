//! Common types shared between state and modules

mod address;

mod module_id;

pub use module_id::{ModuleId, ModuleIdBech32};

mod amount;
mod crypto;
mod error;
mod module_utils;

pub use address::*;
pub use amount::*;
pub use crypto::*;
pub use error::*;
pub use module_utils::*;
pub use sov_state::jmt::Version;
