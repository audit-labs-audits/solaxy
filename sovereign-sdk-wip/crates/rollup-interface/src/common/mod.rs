//! Common types and traits used all throughout the Sovereign SDK.

mod hex_string;
pub mod safe_vec;
mod slot_numbering;

pub use hex_string::*;
pub use safe_vec::SafeVec;
pub use slot_numbering::*;
pub use sov_universal_wallet::schema::safe_string::{SafeString, SizedSafeString};
