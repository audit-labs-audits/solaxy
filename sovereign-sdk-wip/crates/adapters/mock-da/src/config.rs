use std::ops::Range;
use std::time::Duration;

use schemars::JsonSchema;
use sov_rollup_interface::common::HexHash;
use sov_rollup_interface::da::Time;

use crate::storable::layer::StorableMockDaLayer;
use crate::{MockAddress, MockBlock, MockBlockHeader, MockHash};

/// Time in milliseconds to wait for the next block if it is not there yet.
/// How many times wait attempts are done depends on service configuration.
pub const WAIT_ATTEMPT_PAUSE: Duration = Duration::from_millis(10);
/// The max time for the requested block to be produced.
pub const DEFAULT_BLOCK_WAITING_TIME_MS: u64 = 120_000;

pub(crate) const GENESIS_HEADER: MockBlockHeader = MockBlockHeader {
    prev_hash: MockHash([0; 32]),
    hash: MockHash([1; 32]),
    height: 0,
    // 2023-01-01T00:00:00Z
    time: Time::from_millis(1672531200000),
};

pub(crate) const GENESIS_BLOCK: MockBlock = MockBlock {
    header: GENESIS_HEADER,
    batch_blobs: Vec::new(),
    proof_blobs: Vec::new(),
};

/// Configuration for block producing.
#[derive(Debug, Clone, PartialEq, serde::Serialize, serde::Deserialize, JsonSchema)]
#[serde(rename_all = "snake_case")]
pub enum BlockProducingConfig {
    /// Blocks are produced at fixed time intervals, regardless of whether
    /// there are transactions. This means empty blocks may be created.
    Periodic {
        /// The interval, in milliseconds, at which new blocks are produced.
        block_time_ms: u64,
    },

    /// A new block is produced only when a batch blob (but not a proof blob) is submitted.
    /// Each block contains exactly one batch blob and zero or more proof blobs.
    OnBatchSubmit {
        /// The maximum time [`sov_rollup_interface::node::da::DaService::get_block_at`] will wait for a block to become available.
        /// If this timeout elapses, an error is returned.
        /// If set to `None`, [`DEFAULT_BLOCK_WAITING_TIME_MS`] is used.
        block_wait_timeout_ms: Option<u64>,
    },

    /// A new block is produced when either a batch blob or a proof blob is submitted.
    /// Each block contains exactly one blob.
    OnAnySubmit {
        /// The maximum time [`sov_rollup_interface::node::da::DaService::get_block_at`] will wait for a block to become available.
        /// If this timeout elapses, an error is returned.
        /// If set to `None`, [`DEFAULT_BLOCK_WAITING_TIME_MS`] is used.
        block_wait_timeout_ms: Option<u64>,
    },
    /// Blocks are created manually, with no automatic production.
    Manual,
}

/// Defines the behavior of randomization applied to blobs or blocks.
///
/// This configurable behavior determines how blobs are processed and returned to the caller
/// during various stages of the block production process.
/// Randomization may involve reordering, shuffling, skipping, or altering the chain's length,
/// while preserving certain constraints such as finality.
#[derive(Debug, Clone, PartialEq, serde::Deserialize, serde::Serialize, JsonSchema)]
#[serde(rename_all = "snake_case")]
pub enum RandomizationBehaviour {
    /// Blobs inside a single block are rearranged in a random order when read.
    /// This does not affect the boundary between blocks, meaning no blob will
    /// cross its original block's boundary.
    ///
    /// Notes:
    /// - Does not impact how new blocks are produced, and block hashes are not changed.
    /// - Finalized blocks may have their blobs reordered within this mode.
    /// - This does not change the stored order of blobs.
    /// - If randomization is disabled, blobs will be returned in their original order.
    /// - Order guaranteed to be deterministic for each block with the same randomizer configuration.
    OutOfOrderBlobs,
    /// Rewinds the chain to a specific height, chosen randomly between
    /// the most recently finalized block and the current head of the chain.
    ///
    /// This operation adjusts the chain height but maintains finalization constraints.
    Rewind,
    /// Combines blob shuffling with chain height adjustment:
    ///
    /// 1. All non-finalized blobs, including those being added to a new block,
    ///    are shuffled across all blobs that are part of the new chain state.
    /// 2. The chain height is adjusted (rewound or extended) within the constraints
    ///    of the finality window.
    ///
    /// **Constraints**:
    /// - Rewinding can only occur as far back as the finality window allows.
    /// - Extending is not possible if the finality window is already full.
    /// - Rewinding is not triggered if there is only one non-finalized block.
    /// - The specified percentage of blobs (`drop_percent`) is always respected.
    ShuffleAndResize {
        /// Percentage of blobs to be permanently skipped during this process.
        ///
        /// A value of `100` means all non-finalized blobs will be dropped.
        drop_percent: u8,
        /// Range of possible adjustments to the chain head height:
        /// - Negative values represent rewinding the chain length (moving backward in height).
        /// - Positive values represent extending the chain length (adding new blocks).
        /// - This adjustment is constrained by the finality window.
        ///
        /// The actual value is selected by [`crate::storable::layer::Randomizer`] from this range.
        adjust_head_height: Range<i32>,
    },
}

impl RandomizationBehaviour {
    /// Only shuffling without adjusting height of the rollup,
    pub fn only_shuffle(drop_percent: u8) -> Self {
        Self::ShuffleAndResize {
            drop_percent,
            adjust_head_height: 0..1,
        }
    }
}

/// Configuration for randomization applied.
///
/// This struct defines how randomization is performed, including the seed for the randomizer,
/// the timing of chain reorganization, and the specific randomization behavior applied.
#[derive(Debug, Clone, PartialEq, serde::Deserialize, serde::Serialize, JsonSchema)]
pub struct RandomizationConfig {
    /// Seed used by the randomizer to ensure deterministic but randomized behavior.
    pub seed: HexHash,
    /// The interval, in produced blocks, at which chain reorganization may occur.
    /// Applicable for all cases except [`RandomizationBehaviour::OutOfOrderBlobs`],
    /// which does not affect block production.
    ///
    /// For a range `m..n`:
    /// - A reorganization can occur at every `m`-th block produced after the last reorganization.
    /// - A reorganization will definitely occur at or before the `n`-th block produced since the last reorganization.
    ///
    /// Note:
    /// - The interval is counted starting from the height at which the last reorganization happened,
    ///   rather than the current state of the chain.
    /// - This allows the chain to progress consistently within the specified bounds between reorganizations.
    pub reorg_interval: Range<u32>,
    /// Defines the specific behavior of the randomizer during randomization.
    ///
    /// This determines how blobs or blocks are processed, including their ordering,
    /// shuffling, skipping, or potential adjustments affecting the chain.
    pub behaviour: RandomizationBehaviour,
}

/// The configuration for Mock Da.
#[derive(Debug, Clone, serde::Deserialize, serde::Serialize, JsonSchema)]
pub struct MockDaConfig {
    /// Connection string to the database for storing Da Data.
    ///   - "sqlite://demo_data/da.sqlite?mode=rwc"
    ///   - "sqlite::memory:"
    ///   - "postgresql://root:hunter2@aws.amazon.com/mock-da"
    pub connection_string: String,
    /// The address to use to "submit" blobs on the mock da layer.
    pub sender_address: MockAddress,
    /// Defines how many blocks progress to finalization.
    #[serde(default)]
    pub finalization_blocks: u32,
    /// How MockDaService should produce blocks.
    #[serde(default = "default_block_producing")]
    pub block_producing: BlockProducingConfig,
    /// Allow pointing to pre-existing [`StorableMockDaLayer`]
    #[serde(skip)]
    pub da_layer: Option<std::sync::Arc<tokio::sync::RwLock<StorableMockDaLayer>>>,
    /// If specified, [`StorableMockDaLayer`] will add randomization to non-finalized blocks.
    pub randomization: Option<RandomizationConfig>,
}

impl PartialEq for MockDaConfig {
    fn eq(&self, other: &Self) -> bool {
        let basic_eq = self.connection_string == other.connection_string
            && self.sender_address == other.sender_address
            && self.finalization_blocks == other.finalization_blocks
            && self.block_producing == other.block_producing
            && self.randomization == other.randomization;

        // Basic fields are not equal, no need to check da_layer field
        if !basic_eq {
            false
        } else {
            // We can only consider them Eq if `DaLayer` is None in both cases
            self.da_layer.is_none() && other.da_layer.is_none()
        }
    }
}

pub(crate) fn default_block_producing() -> BlockProducingConfig {
    BlockProducingConfig::OnBatchSubmit {
        block_wait_timeout_ms: Some(DEFAULT_BLOCK_WAITING_TIME_MS),
    }
}

impl MockDaConfig {
    /// Create [`MockDaConfig`] with instant finality.
    pub fn instant_with_sender(sender: MockAddress) -> Self {
        MockDaConfig {
            connection_string: Self::sqlite_in_memory(),
            sender_address: sender,
            finalization_blocks: 0,
            block_producing: default_block_producing(),
            da_layer: None,
            randomization: None,
        }
    }

    /// Connection string for in-memory SQLite.
    pub fn sqlite_in_memory() -> String {
        "sqlite::memory:".to_string()
    }

    /// Builds SQlite connection string and checks if a given directory exists.
    pub fn sqlite_in_dir(dir: impl AsRef<std::path::Path>) -> anyhow::Result<String> {
        let path = dir.as_ref();
        if !path.exists() {
            anyhow::bail!("Path {} does no exist", path.display());
        }
        let db_path = path.join("mock_da.sqlite");
        tracing::debug!(path = %db_path.display(), "Opening StorableMockDa");
        Ok(format!("sqlite://{}?mode=rwc", db_path.to_string_lossy()))
    }

    /// Instance of [`MockDaConfig`] that resembles Celestia DA. Batch production is periodic.
    pub fn celestia_like(connection_string: String, sender: MockAddress, seed: HexHash) -> Self {
        MockDaConfig {
            connection_string,
            sender_address: sender,
            finalization_blocks: 0,
            block_producing: BlockProducingConfig::Periodic {
                block_time_ms: 6_000,
            },
            da_layer: None,
            randomization: Some(RandomizationConfig {
                seed,
                // Not really applicable
                reorg_interval: Default::default(),
                // Just to spice things up a bit
                behaviour: RandomizationBehaviour::OutOfOrderBlobs,
            }),
        }
    }

    /// Instance of [`MockDaConfig`] that resembles Solana DA. Batch production is periodic.
    pub fn solana_like(connection_string: String, sender: MockAddress, seed: HexHash) -> Self {
        MockDaConfig {
            connection_string,
            sender_address: sender,
            finalization_blocks: 45,
            block_producing: BlockProducingConfig::Periodic { block_time_ms: 250 },
            da_layer: None,
            randomization: Some(RandomizationConfig {
                seed,
                reorg_interval: 10..20,
                behaviour: RandomizationBehaviour::ShuffleAndResize {
                    drop_percent: 5,
                    adjust_head_height: -10..10,
                },
            }),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn periodic_block_producing() {
        let config_s = r#"
            connection_string = "sqlite:///tmp/mockda.sqlite?mode=rwc"
            sender_address = "0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f"
            finalization_blocks = 5
            [block_producing.periodic]
            block_time_ms = 1_000
        "#;
        let config = toml::from_str::<MockDaConfig>(config_s).unwrap();
        insta::assert_json_snapshot!(config);
    }

    #[test]
    fn manual_block_producing() {
        let config_s = r#"
            connection_string = "sqlite:///tmp/mockda.sqlite?mode=rwc"
            sender_address = "0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f"
            [block_producing.manual]
        "#;
        let config = toml::from_str::<MockDaConfig>(config_s).unwrap();
        insta::assert_json_snapshot!(config);
    }

    #[test]
    fn with_randomization_shuffle_and_resize() {
        let config_s = r#"
            connection_string = "sqlite:///tmp/mockda.sqlite?mode=rwc"
            sender_address = "0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f"
            finalization_blocks = 5
            [block_producing.periodic]
            block_time_ms = 1_000
            [randomization]
            seed = "0x0000000000000000000000000000000000000000000000000000000000000012"
            reorg_interval = [3, 5]
            [randomization.behaviour.shuffle_and_resize]
            drop_percent = 10
            adjust_head_height = [-3, 2]
        "#;
        let config = toml::from_str::<MockDaConfig>(config_s).unwrap();
        insta::assert_json_snapshot!(config);
    }

    #[test]
    fn with_randomization_rewind() {
        let config_s = r#"
            connection_string = "sqlite:///tmp/mockda.sqlite?mode=rwc"
            sender_address = "0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f"
            finalization_blocks = 5
            [block_producing.periodic]
            block_time_ms = 1_000
            [randomization]
            seed = "0x0000000000000000000000000000000000000000000000000000000000000012"
            [randomization.reorg_interval]
            start = 3
            end = 5
            [randomization.behaviour.rewind]
        "#;
        let config = toml::from_str::<MockDaConfig>(config_s).unwrap();
        insta::assert_json_snapshot!(config);
    }
}
