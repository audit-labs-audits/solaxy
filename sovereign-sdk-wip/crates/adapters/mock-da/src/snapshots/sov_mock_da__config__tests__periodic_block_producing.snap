---
source: crates/adapters/mock-da/src/config.rs
expression: config
---
{
  "connection_string": "sqlite:///tmp/mockda.sqlite?mode=rwc",
  "sender_address": "0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f0f",
  "finalization_blocks": 5,
  "block_producing": {
    "periodic": {
      "block_time_ms": 1000
    }
  },
  "randomization": null
}
